crates/solver/src/settlement_submission/submitter/public_mempool_api.rs

```diff
@@ -119,6 +130,77 @@ impl TransactionSubmitting for PublicMempoolApi {
     }
 }
 
+#[derive(Debug, Clone)]
+pub enum SubmissionNode {
+    /// Transactions that are sent to this nodes are expected to be broadcast to the mempool and
+    /// eventually be included in a block.
+    Broadcast(Web3),
+    /// A notification node is an endpoint that is not expected to submit transactions to the
+    /// mempool once a transaction has been received. Its purpose is notifying the node owner that a
+    /// transaction has been submitted.
+    /// In general, there are lower expectations on the availability of this node variant.
+    Notification(Web3),
+}
+
+impl Deref for SubmissionNode {
+    type Target = Web3;
+
+    fn deref(&self) -> &Self::Target {
+        match self {
+            SubmissionNode::Broadcast(web3) => web3,
+            SubmissionNode::Notification(web3) => web3,
+        }
+    }
+}
+
+impl SubmissionNode {
+    pub async fn validated_broadcast_node(
+        ethrpc_configs: &ethrpc::Arguments,
+        http_factory: &HttpClientFactory,
+        url: Url,
+        name: impl ToString,
+        expected_network_id: &String,
+    ) -> Result<Self> {
+        let node = ethrpc::web3(ethrpc_configs, http_factory, &url, name);
+        validate_submission_node(&node, expected_network_id)
+            .await
+            .with_context(|| format!("Validation error for broadcast node {url}"))?;
+        Ok(Self::Broadcast(node))
+    }
+
+    pub async fn from_notification_url(
+        ethrpc_configs: &ethrpc::Arguments,
+        http_factory: &HttpClientFactory,
+        url: Url,
+        name: impl ToString,
+        expected_network_id: &String,
+    ) -> Self {
+        let node = ethrpc::web3(ethrpc_configs, http_factory, &url, name);
+        if let Err(err) = validate_submission_node(&node, expected_network_id).await {
+            tracing::error!("Error validating submission notification node {url}: {err}");
+        }
+        Self::Notification(node)
+    }
```

Comment:
These function seem to do the same (validating that a node belongs to the correct network). Why not have a single function `validate_node()` and then wrap the result in a `SubmissionNode` variant?
If you absolutely want to make sure that you can only create a `SubmissionNode` from a validated node you can put it in a new type like
```rust
struct ValidatedNode(Web3);
fn validate_node() -> Result<ValidatedNode>;
SubmissionNode::new(node: ValidatedNode) -> Self;
```

I think this should allow you to instantiate the nodes in `solver/main.rs:335` with a lot less code.

Reply:
True. I changed this [here](https://github.com/cowprotocol/services/pull/1171/commits/3006c858309b6f11af6d56f6eea18fde382a1167), but I'm not sure I like it more because it puts more logic in `main.rs` that might be useful to know when reading `main`.

Reply:
I had it a bit different in mind. More like this:
```rust
            .map(|(index, url)| {
                let name = format!("broadcast {index}");
                (name, url, true)
            })
            .chain(
                args.transaction_notification_nodes
                    .into_iter()
                    .enumerate()
                    .map(|(index, url)| {
                        let name = format!("notify {index}");
                        (name, url, false)
                    })
            )
            .map(|(name, url, panic_on_error)| async move {
                let node = ethrpc::web3(&args.shared.ethrpc, &http_factory, &url, name);
                if let Err(err) = validate_submission_node(&node, &network_id).await {
                    tracing::error!(
                        "Error validating submission notification node {url}: {err}"
                    );
                    assert!(!panic_on_error);
                }
                Ok(SubmissionNode::Notification(node))
            })
```

Reply:
I agree this is nicer! I implemented this suggestion with some changes [here](https://github.com/cowprotocol/services/pull/1171/commits/7661a3d98b8ee755774a960a0ac8270c6f8f67cb) if you want to take a look.